File: teamscale-client/src/main/java/com/teamscale/client/ProxySystemProperties.java
package com.teamscale.client;

import org.jetbrains.annotations.NotNull;

/**
 * Reads and writes Java system properties values for
 * <ul>
 *     <li>http.proxyHost</li>
 *     <li>http.proxyPort</li>
 *     <li>http.proxyUser</li>
 *     <li>http.proxyPassword</li>
 * </ul>
 * or the corresponding HTTPS counterpart (starting with https instead of http).
 * These values set the proxy server and credentials that should be used later to reach Teamscale.
 */
public class ProxySystemProperties {

	private static final String PROXY_HOST_SYSTEM_PROPERTY = ".proxyHost";
	private static final String PROXY_PORT_SYSTEM_PROPERTY = ".proxyPort";
	private static final String PROXY_USER_SYSTEM_PROPERTY = ".proxyUser";
	private static final String PROXY_PASSWORD_SYSTEM_PROPERTY = ".proxyPassword";

	private final Protocol protocol;

	/**
	 * Indicates, whether the {@link ProxySystemProperties} should return values for the http.proxy* system properties
	 * or the https.proxy* ones
	 */
	public enum Protocol {
		HTTP,
		HTTPS;

		@Override
		public String toString() {
			return name().toLowerCase();
		}
	}

	/**
	 * @return a prefix to the system properties. Used in {@link TeamscaleProxySystemProperties} to differentiate them
	 * from the JVM system properties for proxies.
	 * */
	@NotNull
	protected String getPropertyPrefix() {
		return "";
	}

	/**
	 * @param protocol Indicates, whether the {@link ProxySystemProperties} should use values for the http.proxy* system
	 *                 properties or the https.proxy* ones
	 */
	public ProxySystemProperties(Protocol protocol) {
		this.protocol = protocol;
	}

	/**
	 * Checks whether proxyHost and proxyPort are set
	 */
	public boolean proxyServerIsSet() throws IncorrectPortFormatException {
		return !StringUtils.isEmpty(getProxyHost()) && getProxyPort() > 0;
	}

	/** Checks whether proxyUser and proxyPassword are set */
	public boolean proxyAuthIsSet() {
		return !StringUtils.isEmpty(getProxyUser()) && !StringUtils.isEmpty(getProxyPassword());
	}

	/** @return the http(s).proxyHost system variable */
	public String getProxyHost() {
		return System.getProperty(getProxyHostSystemPropertyName());
	}

	/** @return the http(s).proxyPort system variable. Returns -1 if no or an invalid port was set. */
	public int getProxyPort() throws IncorrectPortFormatException {
		return parsePort(System.getProperty(getProxyPortSystemPropertyName()));
	}

	/** Set the http(s).proxyHost system variable. */
	public void setProxyHost(String proxyHost) {
		System.setProperty(getProxyHostSystemPropertyName(), proxyHost);
	}

	/** @return the name of the system property specifying the proxy host. */
	@NotNull
	protected String getProxyHostSystemPropertyName() {
		return getPropertyPrefix() + protocol + PROXY_HOST_SYSTEM_PROPERTY;
	}

	/** Set the http(s).proxyPort system variable. */
	public void setProxyPort(int proxyPort) {
		setProxyPort(proxyPort + "");
	}

	/** Set the http(s).proxyPort system variable. */
	public void setProxyPort(String proxyPort) {
		System.setProperty(getProxyPortSystemPropertyName(), proxyPort);
	}

	/** Removes the http(s).proxyPort system variable. For testing. */
	/*package*/ void removeProxyPort() {
		System.clearProperty(getProxyPortSystemPropertyName());
	}

	/** @return the name of the system property specifying the proxy port. */
	@NotNull
	protected String getProxyPortSystemPropertyName() {
		return getPropertyPrefix() + protocol + PROXY_PORT_SYSTEM_PROPERTY;
	}

	/** @return the http(s).proxyUser system variable. */
	public String getProxyUser() {
		return System.getProperty(getProxyUserSystemPropertyName());
	}

	/** Set the http(s).proxyUser system variable. */
	public void setProxyUser(String proxyUser) {
		System.setProperty(getProxyUserSystemPropertyName(), proxyUser);
	}

	/** @return the name of the system property specifying the proxy user. */
	@NotNull
	protected String getProxyUserSystemPropertyName() {
		return getPropertyPrefix() + protocol + PROXY_USER_SYSTEM_PROPERTY;
	}

	/** @return the http(s).proxyPassword system variable. */
	public String getProxyPassword() {
		return System.getProperty(getProxyPasswordSystemPropertyName());
	}


	/** Set the http(s).proxyPassword system variable. */
	public void setProxyPassword(String proxyPassword) {
		System.setProperty(getProxyPasswordSystemPropertyName(), proxyPassword);
	}

	/** @return the name of the system property specifying the proxy password. */
	@NotNull
	protected String getProxyPasswordSystemPropertyName() {
		return getPropertyPrefix() + protocol + PROXY_PASSWORD_SYSTEM_PROPERTY;
	}

	/** Exception thrown if the port is in an unknown format and cannot be read from the system properties. */
	public static class IncorrectPortFormatException extends IllegalArgumentException {

		IncorrectPortFormatException(String message, Throwable cause) {
			super(message, cause);
		}
	}

	/** Parses the given port string. Returns -1 if the string is null or not a valid number. */
	private int parsePort(String portString) throws IncorrectPortFormatException {
		if (StringUtils.isEmpty(portString)) {
			return -1;
		}

		try {
			return Integer.parseInt(portString);
		} catch (NumberFormatException e) {
			throw new IncorrectPortFormatException("Could not parse proxy port \"" + portString +
					"\" set via \"" + getProxyPortSystemPropertyName() + "\"", e);
		}
	}
}

File: teamscale-client/src/main/kotlin/com/teamscale/client/ProxySystemProperties.kt
package com.teamscale.client

/**
 * Manages Java system properties for:
 * - http.proxyHost
 * - http.proxyPort
 * - http.proxyUser
 * - http.proxyPassword
 *
 * Or their HTTPS counterparts (https.*).
 * These values set the proxy server and credentials used to reach Teamscale.
 */
open class ProxySystemProperties(private val protocol: Protocol) {

	companion object {
		private const val PROXY_HOST_SYSTEM_PROPERTY = ".proxyHost"
		private const val PROXY_PORT_SYSTEM_PROPERTY = ".proxyPort"
		private const val PROXY_USER_SYSTEM_PROPERTY = ".proxyUser"
		private const val PROXY_PASSWORD_SYSTEM_PROPERTY = ".proxyPassword"
	}

	/**
	 * Enum representing the supported protocols.
	 */
	enum class Protocol {
		HTTP, HTTPS;

		/**
		 * Returns the protocol name in lowercase.
		 */
		override fun toString() = name.lowercase()
	}

	/**
	 * Prefix for the system property keys.
	 * Can be overridden by subclasses to provide a different prefix.
	 */
	protected open val propertyPrefix = ""

	/**
	 * The proxy host system property.
	 */
	var proxyHost: String?
		get() = getProperty(PROXY_HOST_SYSTEM_PROPERTY)
		set(value) {
			setProperty(PROXY_HOST_SYSTEM_PROPERTY, value)
		}

	/**
	 * The proxy port system property.
	 * Must be a positive integer and less than or equal to 65535.
	 */
	var proxyPort: Int
		get() = getProperty(PROXY_PORT_SYSTEM_PROPERTY)?.toIntOrNull() ?: -1
		set(value) {
			check(value > 0) { "Port must be a positive integer" }
			check(value <= 65535) { "Port must be less than or equal to 65535" }
			setProperty(PROXY_PORT_SYSTEM_PROPERTY, value.toString())
		}

	/**
	 * The proxy user system property.
	 */
	var proxyUser: String?
		get() = getProperty(PROXY_USER_SYSTEM_PROPERTY)
		set(value) {
			setProperty(PROXY_USER_SYSTEM_PROPERTY, value)
		}

	/**
	 * The proxy password system property.
	 */
	var proxyPassword: String?
		get() = getProperty(PROXY_PASSWORD_SYSTEM_PROPERTY)
		set(value) {
			setProperty(PROXY_PASSWORD_SYSTEM_PROPERTY, value)
		}

	/**
	 * Retrieves the system property value for the given property key.
	 *
	 * @param property The property key.
	 * @return The property value or null if not set.
	 */
	private fun getProperty(property: String) =
		System.getProperty("$propertyPrefix${protocol}.$property")

	/**
	 * Sets the system property value for the given property key.
	 *
	 * @param property The property key.
	 * @param value The property value to set.
	 */
	private fun setProperty(property: String, value: String?) {
		value?.let {
			check(it.isNotBlank()) { "Value must not be blank" }
			System.setProperty("$propertyPrefix${protocol}.$property", it)
		}
	}

	/**
	 * Checks if the proxy server is set.
	 *
	 * @return True if the proxy host and port are set, false otherwise.
	 */
	fun isProxyServerSet() = !proxyHost.isNullOrEmpty() && proxyPort > 0

	/**
	 * Checks if the proxy authentication is set.
	 *
	 * @return True if the proxy user and password are set, false otherwise.
	 */
	fun isProxyAuthSet() = !proxyUser.isNullOrEmpty() && !proxyPassword.isNullOrEmpty()

	/**
	 * Clears all proxy system properties.
	 */
	fun clear() {
		System.clearProperty("$propertyPrefix${protocol}.$PROXY_HOST_SYSTEM_PROPERTY")
		System.clearProperty("$propertyPrefix${protocol}.$PROXY_PORT_SYSTEM_PROPERTY")
		System.clearProperty("$propertyPrefix${protocol}.$PROXY_USER_SYSTEM_PROPERTY")
		System.clearProperty("$propertyPrefix${protocol}.$PROXY_PASSWORD_SYSTEM_PROPERTY")
	}

	/**
	 * Exception thrown when the port format is incorrect.
	 *
	 * @param message The exception message.
	 * @param cause The cause of the exception.
	 */
	class IncorrectPortFormatException(message: String, cause: Throwable) : IllegalArgumentException(message, cause)
}